### ISSUE TITLE
Code generation gives "AttributeError: 'StatementReturn' object has no attribute 'subnode_expression'" on complex call helper function

### ISSUE THREAD

**Issue by hashuser** [Nuitka/Nuitka#1742]
**Description**
I found Nuitka 1.0.5 has small probability throw ```AttributeError``` during compilation. I've successfully reproduced this error once, but my method isn't quite stable. I will post my method later.

**Environment**
Python 3.10.6
Nuitka 1.0.5
Windows 10 Pro 21H2
DDR4 32GiB

**Command**
```
nuitka --mingw64 --standalone --show-progress --include-module=requests --onefile --output-dir=C:\Users\home\Desktop\Python\1 "C:\Users\Home\PycharmProjects\QYashmak\project\CAIE PastPaper Downloader.py"
```

**My Code**
```Python
import ssl
import requests
import os
import datetime
import multiprocessing
import threading


def get_context():
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.minimum_version = ssl.TLSVersion.TLSv1_3
    context.verify_mode = ssl.CERT_REQUIRED
    context.check_hostname = True
    context.load_default_certs()
    return context


def downloader(url,path):
    header = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.102 Safari/537.36 Edg/104.0.1293.63'}
    while True:
        try:
            r = requests.get(url, headers=header,timeout=20)
            with open(path, 'wb') as file:
                file.write(r.content)
                file.flush()
            break
        except Exception as e:
            if '404' in str(e):
                break


def worker(task):
    threads = []
    path = 'CAIE A-Level PastPapers/'+task[0][0]+'/'+task[0][1]+'/'
    for x in task[1]:
        npath = None
        if 'm' in x[0][:6]:
            npath = path + 'March/'
        elif 's' in x[0][:6]:
            npath = path + 'May June/'
        elif 'w' in x[0][:6]:
            npath = path + 'October November/'
        if npath == None:
            continue
        os.makedirs(npath,exist_ok=True)
        print(x[1])
        if not os.path.exists(npath + x[0]) or not os.path.getsize(npath + x[0]) > 0:
            t = threading.Thread(target=downloader,args=(x[1],npath + x[0],))
            t.start()
            threads.append(t)
    for t in threads:
        t.join()


if __name__ == '__main__':
    os.makedirs('CAIE A-Level PastPapers',exist_ok=True)
    processes = []
    subjects = ['Mathematics - Further (9231)','Chemistry (9701)', 'Biology (9700)', 'Physics (9702)',
                'Mathematics (9709)', 'Economics (9708)']
    for subject in subjects:
        for year in range(2001,int(str(datetime.date.today())[:4])+1):
            folder = []
            header = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.102 Safari/537.36 Edg/104.0.1293.63'}
            content = None
            while True:
                try:
                    r = requests.get('https://papers.gceguide.com/A%20Levels/' + subject.replace(' ','%20') + '/' + str(year) + '/', headers=header,timeout=20)
                    content = r.text
                    break
                except Exception as e:
                    if '404' in str(e):
                        break
            if content != None:
                for x in content.split('<li class=\'file\'>')[1:]:
                    file_name = x[:x.find("' ")].replace("<a href='", "")
                    folder.append((file_name, 'https://papers.gceguide.com/A%20Levels/' + subject.replace(' ','%20') + '/' + str(year) + '/' + file_name))
                p = multiprocessing.Process(target=worker, args=(((subject, str(year)), folder),))
                p.start()
                processes.append(p)
    for p in processes:
        p.join()
    os.system('pause')
```

**Output**
```
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Total memory usage before generating C code: 142.75 MB (149680128 bytes):
C Source Generation: 0.0%|                         | 0/74, __main__Problem with <Node 'EXPRESSION_FUNCTION_CALL'> at internal:1
Problem with <Node 'EXPRESSION_MAKE_TUPLE'> at internal:1
Problem with <Node 'EXPRESSION_OPERATION_BINARY_MOD'> at internal:1
Problem with <Node 'EXPRESSION_BUILTIN_MAKE_EXCEPTION' with {'exception_name': 'TypeError'}> at internal:1
Problem with <Node 'STATEMENT_RAISE_EXCEPTION'> at internal:1
Problem with <Node 'STATEMENT_CONDITIONAL'> at internal:1
Problem with <Node 'STATEMENT_TRY' with {'aborting': False}> at internal:1
Problem with <Node 'STATEMENT_CONDITIONAL'> at internal:1
Problem with <Node 'STATEMENT_TRY' with {'aborting': False}> at internal:1
Nuitka:WARNING: Problem creating function code '__main__$$$function__8_complex_call_helper_star_list_star_dict'.
C Source Generation: 1.4%|▎                        | 1/74, __main__Traceback (most recent call last):
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\__main__.py", line 138, in <module>
    main()
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\__main__.py", line 124, in main
    MainControl.main()
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\MainControl.py", line 930, in main
    result, options = compileTree()
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\MainControl.py", line 792, in compileTree
    makeSourceDirectory()
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\MainControl.py", line 403, in makeSourceDirectory
    source_code = CodeGeneration.generateModuleCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeGeneration.py", line 532, in generateModuleCode
    return _generateModuleCode(module=module, data_filename=data_filename)
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeGeneration.py", line 492, in _generateModuleCode
    function_code, function_decl = generateFunctionBodyCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeGeneration.py", line 437, in generateFunctionBodyCode
    function_code = getFunctionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\FunctionCodes.py", line 602, in getFunctionCode
    return _getFunctionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\FunctionCodes.py", line 643, in _getFunctionCode
    generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 248, in generateStatementSequenceCode
    _generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 237, in _generateStatementSequenceCode
    generateStatementCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 197, in generateStatementCode
    statement_dispatch_dict[statement.kind](
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\TryCodes.py", line 86, in generateTryCode
    generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 248, in generateStatementSequenceCode
    _generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 237, in _generateStatementSequenceCode
    generateStatementCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 197, in generateStatementCode
    statement_dispatch_dict[statement.kind](
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\BranchCodes.py", line 52, in generateBranchCode
    generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 248, in generateStatementSequenceCode
    _generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 237, in _generateStatementSequenceCode
    generateStatementCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 197, in generateStatementCode
    statement_dispatch_dict[statement.kind](
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\TryCodes.py", line 186, in generateTryCode
    generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 248, in generateStatementSequenceCode
    _generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 237, in _generateStatementSequenceCode
    generateStatementCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 197, in generateStatementCode
    statement_dispatch_dict[statement.kind](
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\BranchCodes.py", line 52, in generateBranchCode
    generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 248, in generateStatementSequenceCode
    _generateStatementSequenceCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 237, in _generateStatementSequenceCode
    generateStatementCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 197, in generateStatementCode
    statement_dispatch_dict[statement.kind](
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\RaisingCodes.py", line 91, in generateRaiseCode
    generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 54, in generateExpressionCode
    _generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 101, in _generateExpressionCode
    code_generator(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\ExceptionCodes.py", line 171, in generateBuiltinMakeExceptionCode
    generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 54, in generateExpressionCode
    _generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 101, in _generateExpressionCode
    code_generator(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\OperationCodes.py", line 53, in generateOperationBinaryCode
    _getBinaryOperationCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\OperationCodes.py", line 230, in _getBinaryOperationCode
    generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 54, in generateExpressionCode
    _generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 101, in _generateExpressionCode
    code_generator(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\TupleCodes.py", line 47, in generateTupleCreationCode
    getTupleCreationCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\TupleCodes.py", line 82, in getTupleCreationCode
    helper_code = generateElementCode(elements[0])
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\TupleCodes.py", line 69, in generateElementCode
    generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 54, in generateExpressionCode
    _generateExpressionCode(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\CodeHelpers.py", line 101, in _generateExpressionCode
    code_generator(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\code_generation\FunctionCodes.py", line 764, in generateFunctionCallCode
    needs_check=expression.subnode_function.subnode_function_ref.getFunctionBody().mayRaiseException(
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\nodes\FunctionNodes.py", line 749, in mayRaiseException
    return body is not None and body.mayRaiseException(exception_type)
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\nodes\StatementNodes.py", line 111, in mayRaiseException
    if statement.mayRaiseException(exception_type):
  File "C:\Users\Home\AppData\Local\Programs\Python\Python310\lib\site-packages\nuitka\nodes\ReturnNodes.py", line 47, in mayRaiseException
    return self.subnode_expression.mayRaiseException(exception_type)
AttributeError: 'StatementReturn' object has no attribute 'subnode_expression'
```

**Comment by hashuser** [comment]
I've tested many times and found a method which 100% trigger this error.

1. Copy this code and paste it into an empty .py file
2. Rename this .py file with random valid name
3. Run this command in CMD  ```nuitka --mingw64 --show-progress --include-module=requests --onefile --output-dir=C:\Users\home\Desktop\Python\1 "C:\Users\Home\Desktop\bug.py"```
4. Use Ctrl+C exit Nuitka after about 15 seconds
5. Run this command in CMD ```nuitka --mingw64 --standalone --show-progress --include-module=requests --onefile --output-dir=C:\Users\home\Desktop\Python\1 "C:\Users\Home\Desktop\bug.py"```
6. You will see ```AttributeError: 'StatementReturn' object has no attribute 'subnode_expression'```

**Comment by kayhayen** [comment]
The only point where a prior compilation can make a different to this, would be the caching of bytecode modules, which are once compiled for dependencies and then persisted. That may also affect the function that crashes, because then it may or may not work on that code. 

That would have be a nasty bug, where the statement in question has been released, and deleted the children to make sure reference cycles are broken. The Python garbage collector runs a bit indetermistic at times, but that ought to have nothing to do with it really.

I am now contemplating questions, what happens if a bytecode demoted module, removed from the compilation, has caused these functions to be registeredfor use, to not be used anymore. How they would get freed, and that would demostrate the issue, if they are still going to have them up for code generation, but made unusable really. 

I am not sure about the erractic nature of your finding there, what I have in mind, would not need a pre-condition here. I think the kind of issue I am talking about would go away on next compilation, due to the caching kicking in. That can be disabled however, which would make it every time.

Also as a side note, I really would appreciate `nuitka --version` output rather than manual write ups that are less complete in my book.

**Comment by hashuser** [comment]
```
C:\Users\Home>nuitka --version
1.0.5
Commercial: None
Python: 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Flavor: Unknown
Executable: C:\Users\Home\AppData\Local\Programs\Python\Python310\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 10
```

**Comment by kayhayen** [comment]
I have not experienced a better reproducer so far, but I am sure I also saw it myself once. I will try and craft a case, where through a later optimization, an internal function like this is created, and ends up being unused. What we see here, suggests that it ends up in code generation. I think another way may also that the cached instance became unused, and released, and then another one was not created, but the released one used. I will play around with that once I find the time.

**Comment by kayhayen** [comment]
It appears I lack the time to do this right now, but surely we ought to get to it. What we should do is to have a derived class for these code helpers, and then to *not* release these functions ever.

**Comment by kayhayen** [comment]
I still don't have the reproducer that would enable me to test my theory. However, the devised solution should be workable, however this is delayed still.

**Comment by kayhayen** [comment]
I think I have reproduced the issue. Interestingly, this is indeed related to bytecode caching. When you have bytecode caching, the issue is hidden on a second compilation, except that the bytecode cache is often deactivated by very innocent changes, creating a file in the directory you compile will already do it. 

For 1.4 the cache will be exact and work across different compilation paths. Module existance or non-existance as well as module kind are going to drive it then.

With 1.3 it now happens with clear caches. The bytecode demotion may indeed be to blame, since it stops the use of a function after previous having it used. What is strange though is that the function in question is still used in other places, so it should not have happened.

**Comment by kayhayen** [comment]
I think the complex helper functions which are so called pure function were not visited again, and therefore failed to mark the pure functions they use as used, which allows them to be released. The correction will actually be in a 1.3 hotfix.

**Comment by kayhayen** [comment]
This is solved with 1.4 release now, was already part of 1.3.6 but I forgot to trace it here.
